zanata-war/src/main/java/org/zanata/service/impl/CopyTransWorkFactory.java

```diff
@@ -57,13 +82,374 @@
     @In
     private VersionStateCache versionStateCacheImpl;
 
-    public CopyTransWork createCopyTransWork(HLocale targetLocale,
+    public Integer runCopyTransInNewTx(HLocale targetLocale,
+            HCopyTransOptions options, HDocument document,
+            boolean requireTranslationReview, List<HTextFlow> copyTargets)
+            throws Exception {
+        return runInTransaction(() -> runCopyTrans(targetLocale, options,
```

Comment:
So, this is not about multithreading (or async). Fortunately all the code runs synchronously and sequentially here. It's just that it needs to start a new transaction around the method. The injection won't be a problem because it's running in the same context all the time (as opposed to async or multi-thread). The Runnable can see everything that the method can, and there is no problem with accessing the injected beans because the bean's invoked method (the one calling the runnable) finishes after the runnable.

The posts you reference are useful when we implement Async. For that, we do need a thread executor (already there) as well as manually starting the CDI contexts (already in the demo project). That is a much harder problem to solve.